publisher-sdk/src/main/java/com/criteo/publisher/CriteoInternal.java

```diff
@@ -121,16 +121,16 @@ public void runSafely() {
   }
 
   @Override
-  public void setBidsForAdUnit(Object object, AdUnit adUnit) {
+  public void setBidsForAdUnit(Object object, AdUnit adUnit, @NonNull BidListener bidListener) {
     try {
-      doSetBidsForAdUnit(object, adUnit);
+      doSetBidsForAdUnit(object, adUnit, bidListener);
     } catch (Throwable e) {
       Log.e(TAG, "Internal error while setting bids for adUnit.", e);
```

Comment:
In the worst case, we can consider that not calling the listener === blocking the publisher

Then it is hard to chose what to do here:
If we have an error, then listener could be not invoked at all. A solution is to call the listener in the catch
But calling the listener in the catch may throw another exception => solution: need to catch it again and use `adSuppressed`.
But with async, calling the listener in the catch may end to call the listener twice => solution: wrap listener in an atomic single shot listener.

Just proposing some solution to issues that may not happen. We can simply wait for reaction of publishers.

Reply:
Errors (such as timeout) should trigger the listener as well. The listener will be called on the UI thread. Why would it result in calling the listener twice?

Reply:
By error, I meant `Throwable`. Once could imagine that async task is started and a throwable is thrown then.

Your sample of code for the integration shows that publisher are using a `CountDownLatch` that they will count down everytime the listener is notified.

Here is your sample:
```
private final CountDownLatch latch = new CountDownLatch(TOTAL_COMPETING_SDKS);
criteo.loadAd(publisherAdView, BANNER, new BidCompleteListener() {
     @Override
     public void onBiddingComplete() {
        latch.countDown();
     }
});
competitor.loadAd(publisherAdView, BANNER, new BidCompleteListener() {
     @Override
     public void onBiddingComplete() {
         latch.countDown();
     }
});
 
 
executor.execute(() -> {
    // give a chance to all SDKs to respond, but timeout at 5 secs
    latch.await(5, TimeUnit.SECONDS);
    Handler handler = new Handler(Looper.getMainLooper());
        handler.post(() -> {
          publisherAdView.loadAd();
     });
});
```

Calling twice the listener would make the SDK acts twice.

Still again, it is just a proposition of solution on issues that may be not worth it to handle.

Reply:
The second call (`competitor.loadAd()`) is from a different Ad SDK and does not call our listener. This example was just meant to show how a publisher could block until all Ad SDKs have finished bidding. As far as our listener is concerned, it will only be called once.